### PR TITLE
Add `purge-signatures` subcommand

### LINKED ISSUE #33: Extend casper-db-utils - add ability to remove finality signatures from blocks
This is needed in order to allow faster testing of historical sync to genesis in networks that have no strict finality. Currently, the alternative is to use either testnet or mainnet snapshot, but then the test takes weeks to execute. The new functionality should allow modification of the snapshot and removal of some or all finality signatures in order to simulate "weak" or "no" finality on a given block.

### PR DESCRIPTION
**Pull request by georgepisaltu**
Fixes #33 

This PR adds the `purge-signatures` subcommand. It takes a path to a storage file and lists of block heights for which to purge all signatures or just the required signatures to reach weak but not strict finality.

### CODE REVIEW

File: src/subcommands/purge_signatures/signatures.rs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+        match inverse_map.entry(*weight) {
+            Entry::Vacant(vacant_entry) => {
+                vacant_entry.insert(vec![key]);
+            }
+            Entry::Occupied(mut occupied_entry) => occupied_entry.get_mut().push(key),
+        }
+    }
+    let mut accumulated_sigs: BTreeSet<&PublicKey> = Default::default();
+    let mut accumulated_weight = U512::zero();
+    // Start from the smalles signatures and add them to our pool until weak
```
Comment (rafal-ch) [inline src/subcommands/purge_signatures/signatures.rs]:
```suggestion
    // Start from the smallest signatures and add them to our pool until weak
```

File: src/subcommands/purge_signatures/purge.rs [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        progress_tracker.advance_by(1);
+    }
+    txn.commit()?;
+    Ok(())
+}
+
+pub fn purge_signatures<P: AsRef<Path>>(
+    db_path: P,
+    weak_finality_block_list: Vec<u64>,
+    no_finality_block_list: Vec<u64>,
```
Comment (rafal-ch) [inline src/subcommands/purge_signatures/purge.rs]: A nit:
These two could already be created as `BTreeSet` in `fn run()`, then you can even consider `purge_signatures_for_blocks()` to accept `ExactSizeIterator`.

But that's just a style preference, so feel free to ignore :)

**Review by rafal-ch — APPROVED** [review]
No blockers, good job! :+1:

File: src/subcommands/purge_signatures/purge.rs (hunk 2 above)
Comment (georgepisaltu) [reply src/subcommands/purge_signatures/purge.rs]: Initially I wanted to not do any "computation" logic in `run`, just argument parsing. And a `Vec` is basically just an owned slice, that's what I was going for, but right now it seems a bit weird to have it first collected into a `Vec` and then turned into a `BTreeSet`, so I just did that in `run`. I won't do the `ExactSizeIterator` thing though as I prefer passing in the actual thing. Hope that's ok with you.

File: src/subcommands/purge_signatures/signatures.rs (hunk 1 above)
Comment (georgepisaltu) [reply src/subcommands/purge_signatures/signatures.rs]: Done.

File: src/subcommands/purge_signatures.rs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                .value_name("DB_PATH")
+                .help("Path of the directory with the `storage.lmdb` file."),
+        )
+        .arg(
+            Arg::new(WEAK_FINALITY)
+                .display_order(DisplayOrder::WeakFinality as usize)
+                .required_unless_present(NO_FINALITY)
+                .short('w')
+                .long(WEAK_FINALITY)
+                .takes_value(true)
```
Comment (Fraser999) [inline src/subcommands/purge_signatures.rs:75]: You could have probably used `use_value_delimiter(true)` here instead and that would avoid having to manually parse the multiple values.  Not a big deal though.

File: src/subcommands/purge_signatures/signatures.rs [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+
+    // Store the signature keys sorted by their respective weight.
+    let mut inverse_map: BTreeMap<U512, Vec<&PublicKey>> = BTreeMap::default();
+    for (key, weight) in weights.iter() {
+        match inverse_map.entry(*weight) {
+            Entry::Vacant(vacant_entry) => {
+                vacant_entry.insert(vec![key]);
+            }
+            Entry::Occupied(mut occupied_entry) => occupied_entry.get_mut().push(key),
+        }
```
Comment (Fraser999) [inline src/subcommands/purge_signatures/signatures.rs]:
```suggestion
        inverse_map.entry(*weight).or_default().push(key);
```

File: src/subcommands/purge_signatures/purge.rs [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+            let serialized_signatures = bincode::serialize(&block_signatures)
+                .map_err(|bincode_err| Error::Serialize(*block_hash, bincode_err))?;
+            txn.put(
+                signatures_db,
+                &block_hash,
+                &serialized_signatures,
+                WriteFlags::default(),
+            )?;
+        } else {
+            info!("Couldn't strip signatures for block {block_hash} at height {block_height}");
```
Comment (Fraser999) [inline src/subcommands/purge_signatures/purge.rs]: Maybe worth a `warn!` rather than `info!`?

File: src/subcommands/purge_signatures/tests.rs [hunk 6]
```diff
@@ -0,0 +1,728 @@
...
+            era_weights.era_id(),
+            switch_block_headers[1].1.era_id.successor()
+        );
+        assert_eq!(
+            *era_weights.weights_mut().get(&KEYS[1]).unwrap(),
+            U512::from(100)
+        );
+        assert!(!era_weights.weights_mut().contains_key(&KEYS[0]));
+
+        // Try to update the weights for a inexistent switch block.
```
Comment (Fraser999) [inline src/subcommands/purge_signatures/tests.rs]:
```suggestion
        // Try to update the weights for a nonexistent switch block.
```

File: src/subcommands/purge_signatures/tests.rs [hunk 7]
```diff
@@ -0,0 +1,728 @@
...
+            Err(LmdbError::NotFound) => {}
+            other => panic!("Unexpected search result: {other:?}"),
+        }
+
+        // Block 2 should be the same as before.
+        let block_2_sigs = get_sigs_from_db(&txn, &fixture, &block_headers[1].0);
+        assert!(block_2_sigs.proofs.contains_key(&KEYS[0]));
+        assert!(!block_2_sigs.proofs.contains_key(&KEYS[1]));
+
+        // Block 2 should be the same as before.
```
Comment (Fraser999) [inline src/subcommands/purge_signatures/tests.rs]:
```suggestion
        // Block 3 should be the same as before.
```

File: src/subcommands/purge_signatures/purge.rs [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+#[derive(Default)]
+pub(crate) struct EraWeights {
+    era_id: EraId,
+    weights: BTreeMap<PublicKey, U512>,
+}
+
+impl EraWeights {
+    /// Update the internal structure to hold the validator weights for
+    /// the era given as input.
+    pub(crate) fn refresh_weights_for_era<T: Transaction>(
```
Comment (Fraser999) [inline src/subcommands/purge_signatures/purge.rs:46]: I think this is sufficient for now, but we should ticket updating this method (and any other places in db-utils like this) to take validator-changing upgrades into account.  (We handle this via the `GotBlockAfterUpgradeEraValidators` event in casper-node I think).

For now, maybe the tool could print a warning if the switch block used inside here happens to be the final switch block of a given protocol version?  That would require parsing the subsequent block to check for a different version in it.

File: src/subcommands/purge_signatures.rs (hunk 3 above)
Comment (georgepisaltu) [reply src/subcommands/purge_signatures.rs:75]: I'd like to not do this right now as I'm more confident with manual parsing. This is something we should do at some point in the future though when we finally add CLI usage tests and/or we end up having another parameter which takes multiple values, like a list. Let me know if that's ok with you.

File: src/subcommands/purge_signatures/purge.rs (hunk 8 above)
Comment (georgepisaltu) [reply src/subcommands/purge_signatures/purge.rs:46]: I tried to address this the best I could without starting to query global state for the weights. The tool and the CLI interface is built in such a way that the user knows what blocks they want stripped, so hopefully they don't intentionally choose a switch block for upgrade.

File: src/subcommands/purge_signatures/purge.rs (hunk 5 above)
Comment (georgepisaltu) [reply src/subcommands/purge_signatures/purge.rs]: Done.

File: src/subcommands/purge_signatures/signatures.rs (hunk 4 above)
Comment (georgepisaltu) [reply src/subcommands/purge_signatures/signatures.rs]: Done.

File: src/subcommands/purge_signatures/tests.rs (hunk 6 above)
Comment (georgepisaltu) [reply src/subcommands/purge_signatures/tests.rs]: Done.

File: src/subcommands/purge_signatures/tests.rs (hunk 7 above)
Comment (georgepisaltu) [reply src/subcommands/purge_signatures/tests.rs]: Done.

File: src/subcommands/purge_signatures.rs (hunk 3 above)
Comment (Fraser999) [reply src/subcommands/purge_signatures.rs:75]: Fine by me :)

**Review by Fraser999 — APPROVED** [review]
LGTM - nice one!

**Pull request merged**